### DISCUSSION TITLE
Proper ways for state management in remix?

### DISCUSSION THREAD

**Discussion by muraterzioglu** [remix-run/remix#915] [Q&A] [answered]
Hello fellas 👋 

I new to the remix and I wonder is there any state management library for remix or can I just use React state management libraries like redux?

Thanks a lot!

**Comment by MichaelDeBoey** [comment, marked as answer]
In theory you can use whatever state management library you want, but with Remix it's not necessary to use any of them since you're getting all your data from the server with the `useLoaderData` hook

**Reply by muraterzioglu** [reply]
Thanks for the answer tho! But what should I do if I want to store auth data (like access token or etc)? Should I use localStorage?

**Reply by muraterzioglu** [reply]
That's why I generally use Redux

**Reply by AlexandrLi** [reply]
For auth data, you can use Cookies (https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies)

**Reply by muraterzioglu** [reply]
> For auth data, you can use Cookies (https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies)

Is it okay the store access token in cookies? or how can I be verified that request coming from the user.

**Reply by muraterzioglu** [reply]
Wait after reading documentation one more time I think it's my solution
https://remix.run/docs/en/v1/api/remix#creatememorysessionstorage

Thanks for the answers! @AlexandrLi @MichaelDeBoey

**Reply by sergiodxa** [reply]
@muraterzioglu I don't recommend you to use the `createMemorySessionStorage` for anything else than local development, instead use the `createCookieSessionStorage` or `createSessionStorage` to implement your own using a DB to store the session data.

Why? Because `createMemorySessionStorage` will lose all the session data after a the server is restarted, so every time you deploy everyone will be logged out, if you deploy to a Serverless platform the server will be shutdown after a request (most of the time) so again users will be logged out.

Using the `createCookieSessionStorage` it's really simple, store the session data in a cookie so you don't need to keep it server-side and if the user doesn't clear the cookies manually it will remain logged-in.

Using `createSessionStorage` is a little bit more advanced, but it will allow you to store the session in a DB if you need it, I think `createCookieSessionStorage` is enough for the majority of the apps.

And about your question if it's okey to store an access token in cookies, it's not only okey, it's preferred over using sessionStorage or localStorage in the browser, you can configure a cookie to be  httpOnly so it will no to possible to read it with JS, but any JS code could read the access token in sessionStorage/localStorage, meaning if someone injected JS on your app or if you installed a compromised version of a package and that packages read the session/local storage data it will be able to access your users access token super easy.

So cookies are a better place to store this, and since they are cookies they will be included automatically by the browser on every request to the server, so you don't need to add an Authorization header manually.

**Reply by muraterzioglu** [reply]
Hey, @sergiodxa thanks for your answer! As English is not my native language this explanation is helped me a lot to understand! Thanks again!

**Reply by Himadu2000** [reply]
@sergiodxa As I know using cookies requires user consent according to law, And this is a very terrible UX design. So is there any other way to do this? Thanks in advance.

**Reply by sergiodxa** [reply]
> @sergiodxa As I know using cookies requires user consent according to law, And this is a very terrible UX design. So is there any other way to do this? Thanks in advance.

As far as I know, and I’m no lawyers nor live in a country with that requirement, you don’t need user consent of cookies used to make the website works, only for cookies used for analytics.

But if I’m wrong, you can use any state management available in React.

**Reply by MichaelDeBoey** [reply]
@Himadu2000 As far as I know, you only need to get consent whenever you would be able to strictly identify somebody over multiple sessions when wanting to comply to the GDPR rules.

**Reply by Himadu2000** [reply]
@MichaelDeBoey,

So if I kept a user logged in after a session I need consent?

And that also means I can't keep any user details such as numbers, id, token, name, any details entered on checkout forms like addresses, and I also suppose auto fill details can't be taken without user pressing submit button.

Also I found that some browsers don't delete the cookie after the session in that case what is the best way to remove them? I also assume that we won't need any consent to set user preferences such as theme and settings.

If there is other conditions please let me know.
I really hope a lawyer would have written about this in an article to help us developers.

**Reply by sergiodxa** [reply]
No you only need consent if you plan to track the user to sell ads or for analytics, cookies and user information needed to use the app doesn’t need consent, because the consent is the privacy policy and terms of service.

**Reply by sergiodxa** [reply]
In any case, if you have doubts talk with a lawyer, don’t take law suggestions from someone on the internet

**Reply by ccssmnn** [reply]
> tldr: You don't need permission to use cookies. You need permission to collect personal data. Functional cookies (language preference, dark mode preference) are fine. Auth state might be related to personal data (in your database) and for that you need permission, usually a checkbox in your signup form.

I am not a lawyer, too. But you can read the law [here](https://gdpr-info.eu/). Chapters 2 and 3 are most relevant and quite readable. You don't ask for permission to set cookies. You ask for permission to collect, store and potentially share or sell data. You might use cookies for this, but cookies are not what needs permission. Collection of personal is. Whenever you have a login, you will start collecting user data. Within the EU, you have to inform your users about:

- Why do I collect data?
- What data do I collect?
- Who has access to this data?
- Where and how long do I store the data?
- Can the data be deleted (or disconnected from my identity)?

... among other things. When you sign up for a service, you sign that you agree to the privacy terms of that service and that you have read their privacy statement. After that, using cookies to store authentication states is completely fine.

The cookie banners you see all over the place are asking for permission to track your individual usage data connected to your identity for analytics purposes and ads. Or when you embed other services like YouTube in your website, you are responsible for cookies that YouTube is setting on your website. There are privacy-friendly alternatives for analytics (e.g. [Plausible](https://plausible.io/)).

**Reply by lili21** [reply]
I guess you can only use `useLoaderData` in a `route module` component, right ? than how to use that data in a normal component?

**Reply by ccssmnn** [reply]
@lili21 you can `useFetcher` to make queries on the client side or pass the data down from the `route` to the components as props.

**Reply by sergiodxa** [reply]
> I guess you can only use `useLoaderData` in a `route module` component, right ? than how to use that data in a normal component?

useLoaderData will work on any component, if you call it outside a route component it will give you the data returned by the loader of the root rendering the component, so if you have a component A and it's rendered by the route `routes/profile` the return value of the hook will be the loader data of `routes/profile`, but if you know render the same component on `routes/dashboard` then the result will be the data of the loader of the dashboard.

**Reply by lili21** [reply]
> > I guess you can only use `useLoaderData` in a `route module` component, right ? than how to use that data in a normal component?
> 
> useLoaderData will work on any component, if you call it outside a route component it will give you the data returned by the loader of the root rendering the component, so if you have a component A and it's rendered by the route `routes/profile` the return value of the hook will be the loader data of `routes/profile`, but if you know render the same component on `routes/dashboard` then the result will be the data of the loader of the dashboard.

wow, I didn't know that. that is really impressive~

**Reply by ronyv89** [reply]
What to do with shared data between two components?

**Reply by lili21** [reply]
@ronyv89 is it from server/loader？ then `useLoaderData` would fit it. if not, just the react way, context or third-party state management library.

**Reply by ronyv89** [reply]
No. I have a page which uses 2 components in which there is a form and a form preview, which are in 2 separate component files. So I need to share the data between these 2 components, so without redux, I will have to store the state in the parent component and pass the state variables as props to each component.

**Reply by lili21** [reply]
You are right. Remix doesn't handle client-side state management.

**Reply by barbinbrad** [reply]
If the shared data is from a loader, you can use the useMatches hook. If it’s not just do it the same way you would in any react app.

There’s nothing wrong with passing data as props. But if you’d rather not, I’d recommend using react context instead of redux. If you really want a state management library a lot of people like zustand. There’s a lot less boilerplate.

**Reply by natewu** [reply]
Whats the preferred way to manage lets say a navbar with user information?

Currently I have the root component with a loader and a navbar component that calls the loader and checks if the user is logged in, then it displays username and logout, otherwise login.

But currently if you log out or log in, the prev state stays on the navbar but if i console log the value, it has been updated(correctly logged out or logged in). If I refresh the page then it displays correctly...

Since remix discourages using a state management system, I don't really want to store it there then check that, I would prefer if the loader worked correctly but I am not sure how to fix the displaying issue even though the value has been changed.

**Reply by lili21** [reply]
> root component with a loader

only page/route component can have loader.

> a navbar component that calls the loader

can you show some code? you don't call the loader in a component.

**Reply by ccssmnn** [reply]
@natewu data is different that state. here you can think of data. When logged in, you get user data, otherwise you dont. You use loaders to receive the data and actions to change it. In this case, logout or login are actions. Remix default behavior will refetch data from all active loaders, when an action is called.

I'm not sure how you called the loader here. But you can receive the data in a route component and pass it to where you need it. This data access will provide up to date data.

Remix does not discourage state management. It just covers many use cases where state management systems are currently used: when to get which data from the server, mutations and when to refesh it.

**Reply by sergiodxa** [reply]
> Currently I have the root component with a loader and a navbar component that calls the loader and checks if the user is logged in, then it displays username and logout, otherwise login

This is what I do without issues.

> I am not sure how to fix the displaying issue even though the value has been changed

@natewu are you perhaps storing the loader data in a useState instead of reading directly from the loader data?

Or are you doing the logout with a GET request instead of a POST?

**Reply by natewu** [reply]
@ccssmnn My logout is an action, so currently it's root.tsx(loader) then components/nav.tsx which is is then used within root.tsx and I call the useLoaderData in nav.tsx which gets the data from root.tsx loader.

In nav.tsx I have a logout button that has an action.
In theory it would update the button based on the condition but it does not(if I console log the conditon it indeed has changed).

**Reply by natewu** [reply]
@sergiodxa nope the loader is in root.tsx and I'm calling it from components/nav.tsx which is used in root.tsx

Then logout uses an action that post to the backend /API/logout.

I should also mention I'm using express backend using the adapter so remix and express run off the same server but I don't think that's the problem

**Reply by sergiodxa** [reply]
@natewu try to create a minimal reproduction repo or codesandbox, but the normal behavior is that after an action runs Remix revalidates loaders so your loader data will update to not have the user authenticated anymore

**Reply by natewu** [reply]
@sergiodxa that is true, it seems like atleast in the console log the data has been updated. The action I run then clears the cookie directly in the client side, then the loader gets the data and returns a null auth token so then on the console log it is indeed logged out. Although the UI remains to be "logged in" until I refresh the page.

I'll try to reproduce it again

**Reply by natewu** [reply]
@sergiodxa while i was trying to reproduce the error in a minimal reproduction repo I discovered I was not using an action for the form so I tried to use an action but it was giving some error about formData, turns out formData is null for some reason, I also followed the examples and tutorials and its exactly the same but still nothing in form data. 

I read up on the issue and people say its something to do with remix and express, not exactly sure maybe you could help take a look at that.

[Example repo](https://github.com/natewu/loader-issue) This is the repo for testing, I used bun so the scripts for dev runs bun if you use npm you can change in package.json

I tried codesandbox but it was not working for some reason, the buttons won't click and none of the js is being run not sure why so I just created a repo.

**Reply by lili21** [reply]
@natewu  action and loader only work at route component.  the logout action inside logout component wont' work.

**Reply by lili21** [reply]
and You need to use remix way to submit form data to trigger the revalidate, which will rerun the loader and get the latest data.
otherwise, you need to trigger revalidate manually , by using useRevalidator hook

**Reply by natewu** [reply]
@lili21 I see, I will try using the useRevalidator hook, I was not aware of that.

**Reply by natewu** [reply]
@lili21 thank you, my issue has been solved by the useRevalidator hook for now, not sure if its the best way, but it does work.

**Reply by lili21** [reply]
@natewu the best way is submit the form, instead of fetch the action endpoint.
you can use `Form`, or `useSubmit`, or `useFetcher` to submit the form.

**Reply by lili21** [reply]
this way, it will revalidate automatically.

**Reply by tombohub** [reply]
How do we apply sorting? Do we need to fetch again from the server?

**Reply by sergiodxa** [reply]
@tombohub unless you send the whole dataset to the client, you will need to fetch the sorted list from the server again since the result of the sorting will most likely need data you don't have
